ecl/eclcc/WORKUNITS.rst

```diff
@@ -0,0 +1,862 @@
+how to execute it, and how to format the results.  Understanding the contents of a workunit is a key step to
+understanding how the HPCC system fits together.  This document begins with an overview of the different elements
+in a workunit.  That is followed by a walk-through executing a simple query, with a more detailed description of
+some of the workunit components, to show how they all tie together.
+
+Before looking at the contents of the workunit it is important to understand one of the design goals behind the HPCC
+system.  The ECL system logically splits that code that is needed to execute a query in two.  One the one hand there
+are the algorithms that are used a perform different dataset operations (e.g., sorting, deduping).  The same algorithms are
+used by all the queries that execute on the system.  On the other hand there is the meta-data describing what columns
+are in the datasets, which columns you need to sort by, and the order of operations required by the query.  These
+are typically different for each query.
+
+The workunit only contains information about the meta data i.e., "what to do", while the different engines (hthor,
+roxie, thor) implement the algorithms - "how to do it".  If you look at a workunit for an ECL query that sorts
+a dataset you will not find code to perform the sort in the workunit - you won't even find a call to a sort library
+function - that logical is contained in the engine that executes the query.
+
+One consequence of this split, which can be initially confusing, is that execution continually passes back and forth
+between the generated code and the engines.  By the end of this document you should have a better understanding
+of how the generated code is structured and what part it plays in executing a query.
+
+Note the term "Query" is used as a generic term to cover read-only queries (typically run in roxie) and ETL style
+queries which create lots of persistent datafiles (typically run in Thor).  Also, the term "workunit" is used a bit
+ambiguously.  The dll created from a query is called a workunit (which is static), but "workunit" is also used to
+describe a single execution of a query (which includes the parameters and results).  It should be clear from the
+context which of these is meant.
+
+Contents of a workunit
+======================
+
+A workunit is generated by the ecl compiler, and consists of a single dll/shared object.  That dll contains
+several different elements:
+
+* Various c++ helper classes, and exported factory methods are used to create instances of those classes.
+* An XML resource containing different pieces of information about a workunit, including workflow and graphs.
+* Other user-defined resources included in the manifest.
+
+The workunit dll contains everything that the engines need to execute the query.  When a workunit is executed, key
+elements of the xml information are cloned from the workunit dll and copied into a database (currently dali).  This
```

Comment:
"**D**ali"
Since this is the first mention it should read along the lines of - "...and copied into a database (within the HPCC platform, this database is referred to as 'Dali')